Fixed overlay popups not automatically closing

TomEdwardsEnscape opened the pull request:
Fixes #16559. The issue was introduced by #15596.

## What is the current behavior?
Because overlay popups are hosted within the current window, the code which detects whether the pointer is over the current tooltip produces a false positive.

## Breaking changes
None

## Obsoletions / Deprecations
None

TomEdwardsEnscape committed:
Fixed overlay popups not automatically closing

MrJul commented:
> We have an entire test class (`ToolTipTests_Overlay`) which covers overlay popups. I'm not yet sure why these tests have been passing with such a critical bug in the code.

Indeed! The change looks good, but a matching test (or fix for existing test) would be appreciated, to ensure we don't regress this again.

TomEdwardsEnscape committed:
Fix overlay tooltip tests not actually generating overlay tooltips
Verify popup type whenever we verify that the popup is open

TomEdwardsEnscape commented:
The way in which popups are created changed, so the tests were stuck on default behaviour (a popup). I fixed this and added verification that the popup is of the expected type to prevent future issues.

MrJul commented:
> The way in which popups are created changed, so the tests were stuck on default behaviour (tooltip in a popup). I fixed this and added verification that the tooltip host is of the expected type to prevent future issues.

Thanks.

While I understand the changes and can see that the overlay host is correctly being used, no tests are actually failing if I revert the actual fix in `ToolTipService` while using the new tests, which is a bit perplexing...

(Also I don't know why we weren't using the already correctly configured mock `IWindowPlatform` previously instead of calling the static `MockWindowingPlatform.CreateWindowMock` directly and needing the new workaround - but that's another issue.)

TomEdwardsEnscape committed:
Fixed overlay tooltips not being attached to the visual tree in tests

TomEdwardsEnscape commented:
Excellent point. This secondary issue was caused by overlay tooltips not ever being attached to the visual tree in tests: the layout of the overlay layer is only calculated during rendering, which doesn't happen in headless unit tests. The layer is now manually measured by the overlay test class whenever a tooltip opens, plus we ensure that expected values have been assigned.

I wasn't able to get the configured mock windowing platform object to be used, even if I used `AvaloniaLocator` to request the interface. Not sure why, but the current solution offers a stronger guarantee of correctness than any "clean" approach would so I'm happy with it.

MrJul commented:
> Excellent point. This secondary issue was caused by overlay tooltips not ever being attached to the visual tree in tests: the layout of the overlay layer is only calculated during rendering, which doesn't happen in headless unit tests. The layer is now manually measured by the overlay test class whenever a tooltip opens, plus we ensure that expected values have been assigned.

That makes perfect sense. Thank you for the fix.
